feat(archive): add yearly/monthly archive folder options

fnune opened the pull request:
Fixes #900.
Resolves #8390.

Adds yearly and monthly archive folder organization matching Desktop's behavior. Eliminates the need to manually switch archive folders each year. Does this with an "Archive options" preference with three modes: Single folder, Yearly folders (`Archive/2025`), or Monthly folders (`Archive/2025/03`). Matches Desktop's two-setting approach and uses the same `archiveGranularity` property. Year/month subfolders are created automatically based on message date. Existing accounts default to single folder mode for backward compatibility.

## Design decisions

Two settings: matches Desktop's "which folder?" + "how to organize?" rather than a flat list of "Archive-Yearly", "CustomFolder-Monthly", etc. Avoids combinatorial explosion and works with any base folder.

Automatic subfolder creation: year/month folders are created automatically based on granularity setting, not manually selected. This avoids yearly updates and matches Desktop behavior.

Defaults: new accounts default to yearly (matching Desktop). Existing accounts migrated to single folder to preserve current behavior.

Graceful degradation: if subfolder creation fails, messages are archived to the base archive folder rather than failing the operation entirely.

Uses account's folder path delimiter rather than assuming `/`, since some IMAP servers use `.` or other delimiters.

## Potential follow-up

Issue #905 folder structure preservation, automatic yearly (or yearly + monthly) detection, translations...

## Questions for reviewers

1. Should folder creation failures show user notifications, or is the current approach (graceful fallback to base folder with logging) acceptable?
2. Is `PER_YEAR_ARCHIVE_FOLDERS` the right default for new accounts (matching Desktop), with `SINGLE_ARCHIVE_FOLDER` for migrated accounts to preserve existing behavior?

## Screenshots

<img width="2569" height="1971" alt="Screenshot_20251116_193637" src="https://github.com/user-attachments/assets/2936bb66-6557-4e73-8220-f193277b2e3e" />
<img width="1788" height="1892" alt="Screenshot_20251116_193651" src="https://github.com/user-attachments/assets/48321dad-d65b-4f9d-b7a1-e39cc2335cc0" />
<img width="2142" height="1853" alt="Screenshot_20251116_193921" src="https://github.com/user-attachments/assets/ae17767c-c1f5-42d8-9ba2-f831a6148d48" />
<img width="1819" height="1793" alt="Screenshot_20251116_193950" src="https://github.com/user-attachments/assets/6e7cb3bb-382d-41ba-9066-4595631c31da" />
<img width="1612" height="1003" alt="Screenshot_20251116_194018" src="https://github.com/user-attachments/assets/d77ea85c-1655-426a-a22a-e922c3afca56" />
<img width="1859" height="1302" alt="Screenshot_20251116_194041" src="https://github.com/user-attachments/assets/e85f689f-8b65-4a43-97b9-bc6e73df642d" />
<img width="1806" height="1298" alt="Screenshot_20251116_194109" src="https://github.com/user-attachments/assets/81cf0176-aa95-43ba-9fda-a82548b9113a" />

fnune committed:
feat(settings): add archive granularity options

- Add ArchiveGranularity enum with DEFAULT and MIGRATION_DEFAULT
- Update LegacyAccountDto with archiveGranularity property
- Add database migration v28->v29 for backward compatibility
- Add UI preference with Desktop-matching strings
- Archive options disabled until archive folder selected

fnune committed:
feat(archive): integrate ArchiveFolderResolver

Added ArchiveFolderResolver skeleton with date-based routing stubs.
Updated ArchiveOperations to group messages by destination folder.
Messages are now batched by their resolved archive targets.

fnune committed:
refactor(archive): remove redundant archiveFolderId parameter

Resolver handles all destination logic internally, so the parameter
is no longer needed in archiveMessages and archiveThreads methods.

fnune committed:
feat(archive): implement date-based folder resolution

Resolver now creates date-based subfolders on demand. Messages are routed to yearly (Archive/YYYY) or monthly (Archive/YYYY/MM) subfolders based on their internal/sent date and account granularity setting.

fnune committed:
refactor(archive): use BackendStorage for folder creation

Use BackendStorage.updateFolders() instead of MessageStore.createFolders()
to ensure folders are properly created both locally and remotely. This
matches the pattern used in CreateArchiveFolder use case and ensures
IMAP folder creation is queued for sync.

fnune committed:
test: update MessagingControllerTest for new constructor

Add LegacyAccountDtoBackendStorageFactory mock to test to match
updated MessagingController constructor signature.

fnune committed:
test: add unit tests for ArchiveFolderResolver

Test coverage includes:
- Single folder granularity (no folder creation)
- Yearly folder creation and reuse
- Monthly folder creation and reuse
- Date handling (internal date vs sent date)
- Error handling for folder creation failures
- Month formatting with leading zeros

fnune committed:
refactor: extract thin interfaces for ArchiveFolderResolver

Applied Interface Segregation Principle by extracting two thin
interfaces from thick dependencies:
- FolderIdResolver (2 methods) replaces MessageStoreManager (60+ methods)
- ArchiveFolderCreator (1 method) replaces BackendStorageFactory

Added adapter implementations for production use and created simple
fakes for testing. Tests now use 15-line fakes instead of 200+ line
fakes that implemented entire thick interfaces.

fnune committed:
chore: fix detekt static analysis issues

Resolved detekt warnings by adding explicit locale to String.format,
suppressing intentional generic exception catching, and suppressing
early return count warnings for null safety checks.

fnune committed:
test: rename resolver to testSubject in ArchiveFolderResolverTest

fnune committed:
chore: remove unnecessary comments

fnune committed:
fix: connect archive granularity UI to account settings

Add getString/putString handlers in AccountSettingsDataStore to
properly save and load archive_granularity preference changes.

fnune committed:
test: add StorageMigrationTo29Test for archiveGranularity migration

fnune commented on legacy/core/src/main/java/com/fsck/k9/controller/ArchiveFolderResolver.kt:
```diff
@@ -0,0 +1,84 @@
+            name = subfolderServerId,
+            type = LegacyFolderType.ARCHIVE,
+        )
+        return folderCreator.createFolder(account, folderInfo)
```
I think that after doing this it might be good to resync the folders (basically whatever the button on the screenshot does) because the new folder will not show up on the UI otherwise.

<img width="1599" height="685" alt="Image" src="https://github.com/user-attachments/assets/d5ff0bdd-59e6-47f4-ac28-dbf211bffb63" />

fnune commented on legacy/ui/legacy/src/main/res/values/strings.xml:
```diff
@@ -343,6 +343,11 @@
     <string name="sent_folder_label">Sent folder</string>
     <string name="trash_folder_label">Trash folder</string>
     <string name="archive_folder_label">Archive folder</string>
+    <string name="archive_granularity_label">Archive options</string>
```
I tried to match Thunderbird Desktop here but really there's only one option, so maybe this copy should be thought through once more. This is more stable if the area does get more features, and in that case "Archive options" would be fine, but currently it's a bit meh.

fnune committed:
refactor: rebase and attend to review changes

ryanleesipes commented:
Any chance we can review this and merge so it doesn't go stale? @jbott-tbird

rafaeltonholo commented:
> Any chance we can review this and merge so it doesn't go stale? @jbott-tbird

I will review this PR again later this week.

dani-zilla commented:
We were waiting to request any changes until after the code freeze. It does look like there's a Spotless issue. You can run `./gradlew spotlessCheck` to see what's wrong and either fix it yourself or run `./gradlew spotlessApply` to automatically fix the issues. I do recommend doing the check first to make sure it won't mess anything up.

rafaeltonholo committed:
chore: revert unwanted changes

rafaeltonholo committed:
chore: remove ExperimentalTime usage in ArchiveFolderResolver

The `kotlin.time.Instant` API is no longer experimental, so the `@OptIn(ExperimentalTime::class)` annotation and explicit package prefix are no longer necessary.

rafaeltonholo committed:
fix: add missing folder path delimiter in the folder creation

rafaeltonholo committed:
style: fix import order

rafaeltonholo committed:
refactor: replace `LegacyAccountDtoBackendStorageFactory` with `BackendStorageFactory`

rafaeltonholo committed:
feat(ui): toggle archive granularity based on folder selection

rafaeltonholo commented:
@fnune, @ryanleesipes, I have rebased this branch onto `main` and applied the necessary fixes. Additionally, I added functionality to disable the archive granularity option when the Archive folder is set to None.

@dani-zilla, please review at your convenience.

dani-zilla commented:
I think this looks good. Seems to work as expected. To answer your questions:

>     Should folder creation failures show user notifications, or is the current approach (graceful fallback to base folder with logging) acceptable?

I do like the idea of showing a toast message, similar to the one we'd do if a user tries to move a message that is not yet synced with the server. Small an unobtrusive that tells them there was an issue with the archival process if they selected yearly or monthly, just because they may be upset to find it wasn't going into the expected folder once they finally figure it out. 

>    Is PER_YEAR_ARCHIVE_FOLDERS the right default for new accounts (matching Desktop), with SINGLE_ARCHIVE_FOLDER for migrated accounts to preserve existing behavior?

This I think we should probably have set up to be a single archive folder by default for all new users as well. Just because it's unexpected behavior for most users to start creating archive folders, I would think. Perhaps only the default if they set a specific archive folder, as we do on desktop.

dani-zilla commented:
Speaking with the team, there seem to be a few issues remaining:

- The creation of a server-side archive folder seems to fail
- ... And does so without telling the user
- ... Should we also still allow the creation of a local folder? Ask the user if they want to create it anyway?
- We may want to ensure that yearly is the default archive setting. If this is what is on Desktop and design thinks it's right for mobile as well, we can keep it. Otherwise, we may want to consider other options.
